### PR TITLE
Added a Pick (MiddleClick) event

### PR DESCRIPTION
**Pull request by canitzp**
Like the title says, this pr adds a event if the middle mouse button gets pressed.
It is triggered on the client side only, but the result can be modified so that it gets send to the server.
You can cancel the event so there is no more analyzing for the result or set a new result stack.

### CODE REVIEW

**Comment by shadowfacts** [comment]
I like this idea, but I don't think `MiddleClick` is a good name for it. I'd suggest something like `Pick` or `PickBlock` instead.

**Comment by canitzp** [comment]
A good point, but I think Pick says nothing about it and PickBlock is not correct, cause it works at entities too.

**Comment by shadowfacts** [comment]
Well pick is the terminology used everywhere else in Minecraft (and MCP) for middle-clicking blocks/entities.

**Comment by canitzp** [comment]
You're right and I changed it

**Comment by diesieben07** [comment]
Why does this need Pre and Post? One cancelable event fired from the current place for Post should be enough. And the current implementation does not let you change the pick-result for AIR.

**Comment by canitzp** [comment]
The Pre is to prevent the analyzing of the RayTraceResult, for the case that it is enough for you to cancel the event, otherwise you need the post logic, per example if you want to change the result stack and don't want to simple cancel it.
If you mean with your second point that I only gets fired if you middle click on a block or entity, yes I know, but I see no use for middle click into the air or fluids, since there happens nothing anyway

**Comment by diesieben07** [comment]
I see why you would want to separate it if the "post logic" as you call it (the stuff between the current Pre and Post) was like super complex and expensive to run. But it isn't, especially because this only runs when you do a pick-block, which is not very often (compared to other things).

> but I see no use for middle click into the air or fluids, since there happens nothing anyway

So nobody is allowed to add something there?

**Comment by canitzp** [comment]
I can not get the middle click if it is type MISS, in this case the whole method where my event is fired doesn't get called.
Okay I look into my code and only fire one event instead of Pre and Post

**Comment by canitzp** [comment]
@diesieben07 I collapsed the two events now, but I can't change the AIR click detection, without creating a bigger pr

**Review by diesieben07 — APPROVED** [review]

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java [hunk 1]
```diff
@@ -326,6 +327,42 @@ public LeftClickEmpty(EntityPlayer player, @Nonnull ItemStack stack)
     }
 
     /**
+     * This event is fired on the client side when a player middle clicks while aiming on a block or entity.
+     * There is no update packet for it, you have to send your own when needed.
```
Comment (williewillus) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: make the wording of this match LeftClickEmpty

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java [hunk 2]
```diff
@@ -326,6 +327,42 @@ public LeftClickEmpty(EntityPlayer player, @Nonnull ItemStack stack)
...
+     * You can change the Result ItemStack here or set it to {@link ItemStack#EMPTY} to prevent further updating.
+     * Cancel the event if you want that nothing happens.
```
Comment (williewillus) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: I don't understand the difference between "prevent further updating" and canceling the event.

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java [hunk 3]
```diff
@@ -326,6 +327,42 @@ public LeftClickEmpty(EntityPlayer player, @Nonnull ItemStack stack)
...
+        private RayTraceResult target;
+        private ItemStack currentResult;
+        public Pick(EntityPlayer player, RayTraceResult target, ItemStack currentResult){
+            super(player, EnumHand.MAIN_HAND, target.typeOfHit == RayTraceResult.Type.BLOCK ? target.getBlockPos() : new BlockPos(target.entityHit), target.typeOfHit == RayTraceResult.Type.BLOCK ? target.sideHit : null);
+            this.target = target;
+            this.currentResult = currentResult;
+        }
+
+        /**
+         * @return The same RayTraceResult as {@link net.minecraft.client.Minecraft#objectMouseOver}
```
Comment (williewillus) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: this isn't informative. I'd say
`@return The RayTraceResult representing the object the cursor is over`

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java [hunk 4]
```diff
@@ -326,6 +327,42 @@ public LeftClickEmpty(EntityPlayer player, @Nonnull ItemStack stack)
...
+         */
+        public RayTraceResult getRayTraceResult(){
+            return target;
+        }
+
+        /**
+         * @return The current analyzed result for the RayTraceResult
```
Comment (williewillus) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: drop the word "analyzed" (or drop the javadoc entirely as it should be self explanatory)

**Review by williewillus — CHANGES_REQUESTED** [review]
Your brace formatting is off:
should be `if (foo) { bar; }`, not `if (foo){ bar; }` (missing space)

File: src/main/java/net/minecraftforge/common/ForgeHooks.java [hunk 5]
```diff
@@ -455,6 +455,7 @@ else if (i != -1)
                 }
             }
          */
+
```
Comment (liach) [inline src/main/java/net/minecraftforge/common/ForgeHooks.java]: 👎

File: src/main/java/net/minecraftforge/common/ForgeHooks.java [hunk 6]
```diff
@@ -483,6 +484,13 @@ else if (i != -1)
             result = target.entityHit.getPickedResult(target);
         }
 
+        PlayerInteractEvent.Pick pickEvent = new PlayerInteractEvent.Pick(player, target, result);
+        MinecraftForge.EVENT_BUS.post(pickEvent);
+        if(pickEvent.isCanceled()) {
```
Comment (liach) [inline src/main/java/net/minecraftforge/common/ForgeHooks.java]: According to Forge code style, this should be
```java
if (pickEvent.isCanceled()) {
```

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java [hunk 7]
```diff
@@ -326,6 +327,35 @@ public LeftClickEmpty(EntityPlayer player, @Nonnull ItemStack stack)
...
+     * This event is fired on the client side when a player middle clicks while aiming on a block or entity.
+     * The server is not aware of when the client left clicks empty space, you will need to tell the server yourself.
+     * You can change the Result ItemStack here or set it to {@link ItemStack#EMPTY} or cancel the event to prevent further updating.
+     */
+    @Cancelable
+    public static class Pick extends PlayerInteractEvent
+    {
+        private RayTraceResult target;
+        private ItemStack currentResult;
+        public Pick(EntityPlayer player, RayTraceResult target, ItemStack currentResult) {
```
Comment (liach) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: Left brace `{` should be on a new line. Same for all methods below.
```java
public Pick(EntityPlayer player, RayTraceResult target, ItemStack currentResult)
{
```

**Review by liach — CHANGES_REQUESTED** [review]

**Review by liach — APPROVED** [review]

**Comment by canitzp** [comment]
@williewillus Can you take a look at my last chages and request changes or approve it, so that everybody has accepted it?

**Comment by williewillus** [comment]
could we make canceling the event just set the result to `ItemStack.EMPTY` maybe? 

That way the vanilla logic will catch it to return false and we don't need to patch in the `if (canceled) return false`

idk if that's good forge practice though but it's smaller patch and less duplicate logic

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java [hunk 8]
```diff
@@ -326,6 +327,39 @@ public LeftClickEmpty(EntityPlayer player, @Nonnull ItemStack stack)
     }
 
     /**
+     * This event is fired on the client side when a player middle clicks while aiming on a block or entity.
+     * The server is not aware of when the client left clicks empty space, you will need to tell the server yourself.
```
Comment (williewillus) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: `empty space` -> `block/entity`

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java [hunk 9]
```diff
@@ -326,6 +327,39 @@ public LeftClickEmpty(EntityPlayer player, @Nonnull ItemStack stack)
...
+     * You can change the Result ItemStack here or set it to {@link ItemStack#EMPTY} or cancel the event to prevent further updating.
```
Comment (williewillus) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: `prevent further processing` -> `return nothing from the pick` (or something similar)

File: src/test/java/net/minecraftforge/test/PlayerInteractEventTest.java [hunk 10]
```diff
@@ -152,4 +155,20 @@ public static void interactNormal(PlayerInteractEvent.EntityInteract evt)
             // Applies to both hands
             evt.setCanceled(true);
     }
+
+    @SubscribeEvent
+    public void middleClickPre(PlayerInteractEvent.Pick evt)
```
Comment (williewillus) [inline src/test/java/net/minecraftforge/test/PlayerInteractEventTest.java]: `middleClickPre` -> `pick` or `middleClick`

**Comment by mezz** [comment]
Looks good but can you explain the use of this event? 
What do you need it for, and what ideas do you have that require it?

**Review by williewillus — APPROVED** [review]
looks good to me now

**Comment by canitzp** [comment]
I need this so I can return a custom item block when middle click on a tile entity(with custom nbt), no matter if Vanilla or modded, but only. My only use is for creative mode and not applicable to survival at this time.

**Comment by mezz** [comment]
You can do that for your own blocks by overriding `net.minecraft.block.Block#getPickBlock` without an event.

**Comment by canitzp** [comment]
you said correctly 'at your own' but I want to do this for vanilla and blocks from other mods too

**Comment by mezz** [comment]
I think that sounds questionable, so I'm asking what you have in mind. Can you explain more fully please?

**Comment by canitzp** [comment]
I want to add the funktion that if you pick per example a furnace, instead of receiving the block without nbt, you get it with all the nbt currently available (burntime, inventory,...)
There are other options for triggering it, but I think this is the best way to do it

**Comment by mezz** [comment]
Ok, that makes sense. Thank you for clarifying.

**Review by mezz — CHANGES_REQUESTED** [review]
Please annotate every `ItemStack` field, parameter, and return value in `Pick` as `@Nonnull` so that it follows the convention and helps people use it correctly.

**Comment by KnightMiner** [comment]
Your specific example is a vanilla feature, ctrl pick block copies NBT. But i can imagine this being useful for a vanilla block that does not add the feature, or for a mod added feature.

**Comment by canitzp** [comment]
@mezz Added the annotations

**Review by mezz — APPROVED** [review]

**Comment by mezz** [comment]
Looks good, tested and it works correctly.

File: src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java [hunk 11]
```diff
@@ -326,6 +327,41 @@ public LeftClickEmpty(EntityPlayer player, @Nonnull ItemStack stack)
...
+            this.currentResult = currentResult;
+        }
+
+        public RayTraceResult getRayTraceResult()
+        {
+            return target;
+        }
+
+        @Nonnull
+        public ItemStack getCurrentResult()
```
Comment (mattmess1221) [inline src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: `setStack` may be a better name.

Comment (kashike) [reply src/main/java/net/minecraftforge/event/entity/player/PlayerInteractEvent.java]: I assume you meant `getStack` here, and `setStack` for the below method. :+1:

**Comment by canitzp** [comment]
What do you say @mezz to the name changes? I would prefer that if I change them, I name them getResultStack and setResultStack

**Comment by mezz** [comment]
I think `getResultStack` and `setResultStack` are more clear than the current names.

**Comment by mattmess1221** [comment]
Does the event need to extend `PlayerInteractEvent`? It's not really an interaction. It's also only called on the client afaik.

**Comment by canitzp** [comment]
Since the left click into empty is also their, it has to be there.

**Comment by canitzp** [comment]
And it only get called if a player middle clicks a entity or block -> a interaction

**Comment by mattmess1221** [comment]
I know how the pick block works. I just wasn't sure what else may call that method (nothing else does). I'm saying you should demote the superclass to `PlayerEvent`. It will make the constructor a lot cleaner since you only need to use the player argument (which is always `Minecraft.player` btw.) All the information is provided via the raytraceresult.

Also, I suggest renaming its getter to `getTarget()` since that's what you named the field.

**Comment by williewillus** [comment]
No, we can't rename methods in PlayerInteractEvent. that will break every single mod using it with the current names. 

if we are subclassing PlayerInteractEvent, the stack passed to the superclass should just be the one the player is holding in their main hand, but I suggest not subclassing PIE and just put it separately.

**Comment by canitzp** [comment]
@williewillus I only renamed the methods related to my pr and I don't pass a ItemStack to PIE, I save them in a seperated variable

**Comment by williewillus** [comment]
Edit: I misread

**Comment by williewillus** [comment]
I still don't think you should subclass PIE though

**Comment by canitzp** [comment]
In my eyes that a player clicks a button on a Block or Entity, he interacts with it and so I decided to subclass PIE instead of a other one.
@mezz what do you say about this

**Comment by williewillus** [comment]
I'd prefer it to be separate, but it's fine like this too :P

**Comment by mattmess1221** [comment]
To me, pick block is *not* an interaction. It doesn't change the world nor is it required to inform the server. It only changes your hotbar and tells the server what item you just picked in creative.

Extending `PlayerEvent` is enough, and everything needed can be gotten from the `RayTraceResult`.

**Comment by canitzp** [comment]
Left click empty one subclass above mine is also a client only event, but it is a PIE. Why? Cause the player does interact with a thing (or in this case with nothing->air/fluids) and my pick event is the same thing, with another trigger (if it is a block or entity) and another mouse button. Why should this be not a PIE?

**Comment by mattmess1221** [comment]
That event is still an interaction because the player is "interacting" with air. It's just not a vanilla interaction.

**Comment by canitzp** [comment]
And with middle click the player interact with things too. Only cause the player isn't using it as often as the left or right click  interaction it is still one.

**Comment by gigaherz** [comment]
My opinion on this is divided. On one side "picking a block" sounds a lot like an interaction, but it doesn't actually change the world the world per se:

- It does not activate (run code on) an item
- It does not activate a block
- It does not activate an entity

What it does is just say "I the hotbar slot is now using this item", which happens to be selected based on the thing you have in your crosshairs. Which could be argued to be a type of interaction....

**Comment by canitzp** [comment]
It is possible to let run things with it, after this pr and you can set the stack that get send to the server and change if, after the interaction.
This definition I found online "the action or influence of things on one another", in my mind this sounds like as soon as the player do a thing like click on a other thing, it's an interaction, no matter if server or client side or if it trigger anything

**Comment by mattmess1221** [comment]
I think this would be more of an observation. Nothing is being changed or messed with. We're simply looking at the block/entity and copying it.

**Comment by williewillus** [comment]
After thinking about it a bit more, I actually backtrack. All the other clicks the player does in world (left/right on blocks, entities, and air) are routed through PIE (for better or for worse, you can yell at me later about that :P), so it makes sense for this middle click to be with PIE as well, if only for consistency sake

**Comment by mattmess1221** [comment]
I guess that does make sense. Some modder could be creative and allow middle click to have special code, essentially making it an interact.

I just don't like how many ternary operators are in the constructor. It might be nice to do all the logic outside the event so it can be cleaner.

**Comment by LexManos** [comment]
Seems you have all missed the most important part, explaining why you need this.
As @KnightMiner  said, his want is a vanilla mechanic. So... Why?

**Comment by canitzp** [comment]
The ctrl middle click way only copies the NBT that is written trough writeNBT, but I want to change this so it also writes per example the inventory, stored fluids or other capability related things and I want to not only apply this to vanilla blocks.

**Comment by LexManos** [comment]
None of that information is available client side...

**Comment by canitzp** [comment]
That doesn't mean that I can't send a packet, per example cancel the event and send my own instead of the vanilla one, shouldn't be such a problem

**Comment by mezz** [comment]
When reviewing this with Lex, it seems like the main objection is that middle click is handled on the client when it really should be on the server, which is a vanilla problem. Adding a new client-side event for something that should be on the server seems objectionable.
I'll talk to him again about what the solution here should be.

In the meantime please resolve the merge conflicts.

**Comment by mezz** [comment]
Talked to Lex and we're unfortunately rejecting this because it's a bit too much of a mess for a small gain with the current vanilla implementation.
This type of functionality should be possible with a custom item or something else, just not with middle click until we get a vanilla server-side implementation of it.

**Pull request closed without merging**